File: clients/python/media_storage/caching_proxy.py
"""
media_storage.caching_proxy
===========================

Provides an implementation for a client that works with a local caching proxy.

Usage
-----

This module is not meant to be used externally; relevant objects are exported
to the package level.

Legal
-----

This file is part of the LGPLed Python client of the media-storage project.
This package is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published
by the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU General Public License and
GNU Lesser General Public License along with this program. If not, see
<http://www.gnu.org/licenses/>.
 
(C) Neil Tallim, 2011
"""
import json

import interfaces
import common
import compression
import tempfile

_TEMPFILE_SIZE = 256 * 1024 #Keep reasonable-sized tempfiles in memory

class CachingProxyClient(interfaces.RetrievalConstruct):
    """
    Implements the retrieval mechanisms of the client specification such that all interaction takes
    place with a server running on the same host as the client, with that server being responsbible
    for maintaining a local cache of requested data, to reduce overall network load and increase
    response times for high-demand resources.
    
    All operations performed by an instance of this class are atomic and instantaneous, but they
    incur some additional local server load, meaning that it is a poor choice for data that is
    accessed infrequently, but a good choice for templates and attachments that are sent out
    repeatedly during campaigns.
    """
    def __init__(self, server_host, server_port, proxy_port):
        """
        `server_host` and `server_port` indicate the address of the server to be used for
        operations; the port must be an integer and the host may be an IP or name.
        
        `proxy_port` is the port on which the caching proxy is listening on the 'localhost'
        interface.
        """
        self._server_host = server_host
        self._server_port = server_port
        self._proxy = 'http://localhost:%(port)i/' % {
         'port': proxy_port,
        }
        
    def get(self, uid, read_key, output_file=None, decompress_on_server=False, timeout=5.0):
        """
        Retrieves the requested data from the local proxy, returning its MIME and the decompressed
        content as an open filehandle in a tuple.
        
        `output_file` is the path of the file to which the response is written; an anonymous
        tempfile is used by default. If supplied, the caller is responsble for cleaning it up.
        
        `decompress_on_server` is ignored; the caching proxy is responsible for that.
        
        `timeout` defaults to 5.0s.
        
        All other arguments are the same as in ``media_storage.interfaces.RetrievalConstruct.get``.
        """
        request = common.assemble_request(self._proxy + common.SERVER_GET, {
         'uid': uid,
         'keys': {
          'read': read_key,
         },
         'proxy': {
          'server': {
           'host': self._server_host,
           'port': self._server_port,
          },
         },
        })
        (properties, response) = common.send_request(request, timeout=timeout)
        
        if not output_file:
            output = tempfile.SpooledTemporaryFile(_TEMPFILE_SIZE)
        else:
            output = open(output_file, 'wb')
        output.write(response)
        output.seek(0)
        
        return (properties.get(common.PROPERTY_CONTENT_TYPE), output)
        
    def describe(self, uid, read_key, timeout=2.5):
        """
        Retrieves the requested record from the local proxy as a dictionary.
        
        `timeout` defaults to 2.5s.
        
        All other arguments are the same as in
        ``media_storage.interfaces.ControlConstruct.describe``.
        """
        request = common.assemble_request(self._proxy + common.SERVER_DESCRIBE, {
         'uid': uid,
         'keys': {
          'read': read_key,
         },
         'proxy': {
          'server': {
           'host': self._server_host,
           'port': self._server_port,
          },
         },
        })
        (properties, response) = common.send_request(request, timeout=timeout)
        return json.loads(response)
        

File: clients/c-sharp/CachingProxy.cs
//
//  Author:
//       Neil Tallim <flan@uguu.ca>
// 
//  Copyright (c) 2012 Ivrnet, inc.
// 
//  This library is free software; you can redistribute it and/or modify
//  it under the terms of the GNU Lesser General Public License as
//  published by the Free Software Foundation; either version 2.1 of the
//  License, or (at your option) any later version.
// 
//  This library is distributed in the hope that it will be useful, but
//  WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
//  Lesser General Public License for more details.
// 
//  You should have received a copy of the GNU Lesser General Public
//  License along with this library; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

namespace MediaStorage{
    /// <summary>
    /// A means of accessing a service running on localhost that pulls files from a central server and
    /// caches them locally for frequent re-use. Data is made available synchronously.
    /// </summary>
    public class CachingProxy : AbstractBaseClient, Interfaces.RetrievalConstruct{
        /// <summary>
        /// The address of the media-storage proxy.
        /// </summary>
        private string proxy;
		
        /// <summary>
        /// Initializes a new instance of the <see cref="CachingProxy"/> class.
        /// </summary>
        /// <param name='server'>
        /// The details of the media-storage server.
        /// </param>
        /// <param name='proxy_port'>
        /// The port on which the media-storage proxy is listening.
        /// </param>
        public CachingProxy(Server server, ushort proxy_port){
            this.server = server;
            this.proxy = string.Format("http://localhost:{0}/", proxy_port);
        }

        /// <summary>
        /// Retrieves the requested data from the cache.
        /// </summary>
        /// <returns>
        /// Returns the content's MIME and the decompressed data as a stream (optionally that
        /// supplied as <c>output_file</c>), along with the length of the content in bytes.
        /// </returns>
        /// <param name='uid'>
        /// The UID of the record to be retrieved.
        /// </param>
        /// <param name='read_key'>
        /// A token that grants permission to read the record.
        /// </param>
        /// <param name='output_file'>
        /// An optional stream into which retrieved content may be written; if <c>null</c>, the
        /// default, an on-disk, self-cleaning tempfile, is used instead.
        /// </param>
        /// <param name='decompress_on_server'>
        /// Ignored, since this decision is up to the proxy.
        /// </param>
        /// <param name='timeout'>
        /// The number of seconds to wait for a response; defaults to 5.
        /// </param>
        /// <exception cref="System.Exception">
        /// Some unknown problem occurred.
        /// </exception>
        /// <exception cref="Exceptions.ProtocolError">
        /// A problem occurred related to the transport protocol.
        /// </exception>
        /// <exception cref="Exceptions.UrlError">
        /// A problem occurred related to the network environment.
        /// </exception>
        /// <exception cref="Exceptions.NotFoundError">
        /// The requested record was not found.
        /// </exception>
        /// <exception cref="Exceptions.NotAuthorisedError">
        /// The requested record was not accessible with the given credentials.
        /// </exception>
        public Structures.Internal.Content Get(string uid, string read_key, System.IO.Stream output_file=null, bool decompress_on_server=false, float timeout=5.0f){
            Jayrock.Json.JsonObject get_json = new Jayrock.Json.JsonObject();
            get_json.Add("uid", uid);

            Jayrock.Json.JsonObject keys = new Jayrock.Json.JsonObject();
            keys.Add("read", read_key);
            get_json.Add("keys", keys);

            Jayrock.Json.JsonObject proxy = new Jayrock.Json.JsonObject();
            proxy.Add("server", this.server.ToDictionary());
            get_json.Add("proxy", proxy);

            System.Net.HttpWebRequest request = Libraries.Communication.AssembleRequest(this.proxy + Libraries.Communication.SERVER_GET, get_json);
            Libraries.Communication.Response response = Libraries.Communication.SendRequest(request, timeout:timeout);

            Structures.Internal.Content content = new Structures.Internal.Content();
            if(output_file != null){
                content.Data = output_file;
            }else{
                content.Data = new Libraries.TempFileStream();
            }
            response.Data.CopyTo(content.Data);
            content.Length = content.Data.Length;
            content.Data.Seek(0, System.IO.SeekOrigin.Begin);
            content.Mime = (string)response.Properties[Libraries.Communication.PROPERTY_CONTENT_TYPE];

            return content;
        }

        /// <summary>
        /// Retrieves details about the requested record from the cache.
        ///
        /// The provided details may not be entirely current, since cached copies will be used when possible.
        /// </summary>
        /// <param name='uid'>
        /// The UID of the record to be read.
        /// </param>
        /// <param name='read_key'>
        /// A token that grants permission to read the record.
        /// </param>
        /// <param name='timeout'>
        /// The number of seconds to wait for a response; defaults to 2.5.
        /// </param>
        /// <exception cref="System.Exception">
        /// Some unknown problem occurred.
        /// </exception>
        /// <exception cref="Exceptions.ProtocolError">
        /// A problem occurred related to the transport protocol.
        /// </exception>
        /// <exception cref="Exceptions.UrlError">
        /// A problem occurred related to the network environment.
        /// </exception>
        /// <exception cref="Exceptions.NotFoundError">
        /// The requested record was not found.
        /// </exception>
        /// <exception cref="Exceptions.NotAuthorisedError">
        /// The requested record was not accessible with the given credentials.
        /// </exception>
        public Structures.Internal.Description Describe(string uid, string read_key, float timeout=2.5f){
            Jayrock.Json.JsonObject describe = new Jayrock.Json.JsonObject();
            describe.Add("uid", uid);

            Jayrock.Json.JsonObject keys = new Jayrock.Json.JsonObject();
            keys.Add("read", read_key);
            describe.Add("keys", keys);

            Jayrock.Json.JsonObject proxy = new Jayrock.Json.JsonObject();
            proxy.Add("server", this.server.ToDictionary());
            describe.Add("proxy", proxy);

            System.Net.HttpWebRequest request = Libraries.Communication.AssembleRequest(this.proxy + Libraries.Communication.SERVER_DESCRIBE, describe);
            return new Structures.Internal.Description(Libraries.Communication.SendRequest(request, timeout:timeout).ToDictionary());
        }
		
		/// <summary>
		/// Provides the address of the server to access for queries.
		/// </summary>
		internal override string GetServer(){
			return this.proxy;
		}
    }
}
